sql/src/main/java/io/crate/metadata/settings/SessionSettings.java

```diff
@@ -0,0 +1,120 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.metadata.settings;
+
+import io.crate.metadata.SearchPath;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.io.stream.Writeable;
+
+import javax.annotation.Nullable;
+import java.io.IOException;
+import java.util.Objects;
+
+public final class SessionSettings implements Writeable {
+
+    private final String userName;
+
+    private final SearchPath searchPath;
+
+    @Nullable
+    private final Boolean semiJoinsRewriteEnabled;
```

Comment:
Why can these be `null` ? And how does `null` behave?

Reply:
There are mandatory settings (eg. search_path) and optional (the ones above).
The rationale is that we always want to serialize the former whether we will decide if we also need to serialise the latter, hence the 2 separate constructors and the use of primitive types to make this explicit.

The decision will be based on the new Function Feature (separate PR will follow up on that).
So far, these are not used. I will check later on how this should behave.

Reply:
serialization the additional `null` value is more expensive than keeping it at `true` and `false`. Regardless whether the value is required or not.

With "how does `null` behave" I meant to imply that the semantics are not clear: Is `null` a legitimate value for `semiJoinsRewriteEnabled` or is it "absent" / "not known". 

So In this case I'd go with the primitive type given that it is cheaper anyway.

Reply:
@mfussenegger pushed fixup

I was trying to prepare for the "optimised" context serialisation, thus I have to agree that with the current state (always include `searchPath` and choose whether to also serialise the other 2  boolean Settings), we don’t benefit much from that.

I have dropped the @nullable and used primitive types instead.